File: prob_blast.py
import os
import time
import argparse
import pickle as pkl

from concurrent.futures import ProcessPoolExecutor
from functools import partial
# from Bio.Blast import NCBIWWW, NCBIXML
from pprint import pprint

import utils
from utils.params import params


def main():
    parser = argparse.ArgumentParser(description='Probabilistic BLAST')
    parser.add_argument('--config', '-c', type=str, default='config.py', help='[string] Path of config file')
    args = parser.parse_args()
    params.configure(args.config)

    d = utils.get_prob_seq(params.d, params.d_conf, params.S)
    q = utils.seq_from_fasta(params.q)

    prob_blast(
        q, d,
        params.w, params.S,
        params.hit_thres, params.delta,
        params.hsp_thres, params.e_thres,
    )

    # result_handle = NCBIWWW.qblast('blastn', 'nt', open('query.py', 'r').read())
    # result_handle = NCBIWWW.qblast('blastn', 'nt', q)
    # with open('res.xml', 'w') as fout:
    # fout.write(result_handle.read())

    # blast_recs = NCBIXML.parse(open('res.xml', 'r'))
    # for rec in blast_recs:
    #     pprint(rec.__dict__)


def prob_blast(q, d, w, S, hit_thres, delta, hsp_thres, e_thres):
    '''
    Return local alignments
    '''
    alignments = []
    index = prob_index_table(d, w, S, hit_thres)

    hsps = prob_extend(q, d, w, index, hit_thres, delta, hsp_thres, e_thres)
    prob_extend_gap(hsps, q, d, w, hit_thres, hsp_thres, delta)

    return alignments


def prob_index_table(d, w, S, hit_thres):
    '''
    Index database if doesn't already exist at path
    Return index table as dict
    '''
    start = time.time()
    path = f'data/processed/prob_index_table.w{w}.hit_thres{hit_thres}.pkl'

    if os.path.exists(path):
        return pkl.load(open(path, 'rb'))

    print(f'Indexing database with word size {w}...')

    seeds = utils.gen_seeds(S, w)

    with ProcessPoolExecutor() as executor:
        index = dict(filter(
            lambda x: x,
            executor.map(
                partial(_prob_index_table, d, w, S, hit_thres),
                seeds
            )
        ))
    pkl.dump(index, open(path, 'wb'))

    print(f'Time elapsed: {time.time() - start:.2f}s')
    return index


def _prob_index_table(d, w, S, hit_thres, seed):
    seed = ''.join(seed)
    indices = utils.prob_find(seed, d, w, S, hit_thres)
    return (seed, indices) if indices else ()


def prob_extend(q, d, w, S, index, hit_thres, delta, hsp_thres, e_thres):
    '''
    Ungapped extension
    '''
    hsps = []

    for q_idx in range(len(q) - w + 1):
        seed = q[q_idx:q_idx + w]

        try:  # In case seed not in index table
            for d_idx in index[seed]:
                hsp_left, score_left = prob_left(q_idx, d_idx, q, d, w, S, hit_thres, delta)
                hsp_right, score_right = prob_right(q_idx, d_idx, q, d, w, S, hit_thres, delta)

                if score_left == score_right == float('-inf'):
                    continue

                elif score_left == float('-inf'):
                    hsp_score = score_right

                elif score_right == float('-inf'):
                    hsp_score = score_left

                else:
                    hsp_score = score_right + score_left

                # Need to figure out e-value here
                # Is this right?
                # if utils.e_value(hsp_score) < e_thres:
                #     hsps.append((hsp_left, hsp_right))

                # Or this: hsp_score and ignore e_value
                if hsp_score > hsp_thres:
                    hsps.append((hsp_left, hsp_right))

        except KeyError:
            continue

    return hsps


def prob_left(q_idx, d_idx, q, d, w, S, hit_thres, delta):
    return _prob_extend(q_idx, d_idx, q, d, w, S, hit_thres, delta, step=-1)


def prob_right(q_idx, d_idx, q, d, w, S, hit_thres, delta):
    return _prob_extend(q_idx, d_idx, q, d, w, S, hit_thres, delta, step=1)


def _prob_extend(q_idx, d_idx, q, d, w, S, hit_thres, delta, step):
    max_score = float('-inf')
    max_q_idx = q_idx
    max_d_idx = d_idx
    score = 0

    while 0 < q_idx < len(q) - 1 and 0 < d_idx < d.shape[1] - 1:
        q_idx += step
        d_idx += step

        score += d[S.index(q[q_idx]), d_idx] - hit_thres
        if score > max_score:
            max_score = score
            max_q_idx = q_idx
            max_d_idx = d_idx

        elif max_score - score > 10:
            break

    return (max_q_idx, max_d_idx), max_score


def prob_extend_gap(highest_scoring_pairs, q, d, w, hit_thres, hsp_thres, delta):
    '''
    Gapped extension
    '''

    b = 0.7  # base gap cost
    e = 0.2  # gap extension cost

    hsa = []  # Highest Scoring Alignments

    for hsp in highest_scoring_pairs:
        l_index = hsp[0]
        r_index = hsp[1]

        q_index = q[l_index, r_index]
        # Perform Left gapped extension
        if r_index != 0:
            score_right = prob_right_g(l_index, r_index, q_index, d, w, hit_thres, hsp_thres, delta, b, e)

        if r_index != d.length - 1:
            score_left = prob_left_g(l_index, r_index, q_index, d, w, hit_thres, hsp_thres, delta, b, e)

        hsa_score = score_right + score_left

        # hsa.append((leftIndex, rightIndex, score))

    pass


def prob_right_g(l_index, r_index, q_idx, d_idx, q, d, w, S, hit_thres, delta, b, e):
    m = r_index - l_index + 1
    n = d.shape[1] - 1

    nw_table = [m][n]

    nw_table[0][0] = 0

    i = 1
    j = 1

    while i < m:
        nw_table[i][0] = nw_table[i - 1][0] - b
        i += 1
        

    while j < n:
        nw_table[0][j] = nw_table[0][j - 1] - b
        j+= 1

    for i in range(m):
        for j in range(n):
            '''
            1) perform column row comparison by comparing the character in the query row with the associated probability for the same character in the probablisitc table for the current column position
            2) the hsp right index + column number + 1 is the index for the database
            3) hsp right index + row number + 1 is the index for the query
            4) if the comparison meets the threshold, use the same normal NW scoring scheme?
            5) complex backtrace to generate score or just to note the path? And then perhaps forward trace through the path, generating the score and keeping track of whether or not the current score passes the delta threshold
            6) when delta threshold met or when end of row or column range, then stop
            7) return HSA alignment? How to return exact alignment like BLAST website does?
            '''
            return

    return 0


def prob_left_g(l_index, r_index, q_idx, d_idx, q, d, w, S, hit_thres, delta):
    return 0


if __name__ == '__main__':
    main()

File: prob_blast.go
package main

import (
	"os"
	"fmt"
	"math"
	"sort"
	"github.com/kr/pretty"
	"github.com/Mal-Jovi/561_Project/utils"
	"github.com/Mal-Jovi/561_Project/utils/indexing"
	"github.com/Mal-Jovi/561_Project/utils/ungapped_extension"
	"github.com/Mal-Jovi/561_Project/utils/gapped_extension"
	. "github.com/Mal-Jovi/561_Project/utils/structs"
)

func main() {
	// Default config file path config.json
	config := "config.json"
	if len(os.Args[1:]) == 1 {
		// Parse command line argument for config file path
		config = os.Args[1]

	} else if len(os.Args[1:]) > 1 {
		// > 1 command line arguments are invalid. Stop
		fmt.Println("Invalid arguments. Usage: ./prob_blast [config file]")
		return
	}

	// Parse parameters from config file
	params := utils.ParamsFromJson(&config)
	pretty.Println(params)

	// Load query sequence q, probabilistic sequence d
	q := utils.SeqFromFasta(params.Q)
	d := utils.GetProbSeq(params.D, utils.GetDConf(params.DConf), &params.S)

	// Run probabilistic BLAST
	alignments := prob_blast(q, d, params)
	pretty.Println(*alignments)
	utils.SaveAlignments(alignments, params)
}

// prob_blast
//
// Probabilistic BLAST
//
func prob_blast(q * string, d *[][] float64, params *Params) *[]*Alignment {
	index := prob_index_table(d, params)
	hsps := prob_extend(q, d, index, params)
	alignments := prob_extend_gap(q, d, hsps, params)
	
	return alignments
}

// prob_index_table
//
// Index database, then export the index table
//
func prob_index_table(d *[][]float64, params *Params) *map[string][]int {
	path := fmt.Sprintf("data/processed/prob_index_table.w.%d.hit_thres.%.2f.json", params.W, params.HitThres)

	if _, err := os.Stat(path); !os.IsNotExist(err) {
		return utils.IndexFromJson(&path)
	}

	fmt.Println("Indexing database with word size", params.W, "and hit_thres", params.HitThres, "...")

	index := make(map[string][]int)
	seeds := indexing.GenSeeds(&params.S, params.W)

	// Concurrently index database using num_cores cores
	var start int
	var end int

	for i := 0; i < params.NumCores - 1; i++ {
		start = i * (len(seeds) / params.NumCores) // Start index
		end = (i + 1) * (len(seeds) / params.NumCores) // End index

		// Spawn go routine for concurrency
		go indexing.ProbIndexTable(&index, start, end, &seeds, d, params)
	}
	start = (params.NumCores - 1) * (len(seeds) / 4)
	end = len(seeds)
	indexing.ProbIndexTable(&index, start, end, &seeds, d, params)
	
	// Save index table to path
	utils.ExportToJson(&index, &path)
	fmt.Println("Saved index to", path)

	return &index
}

// prob_extend
//
// Ungapped extension. Return list of high scoring pairs
//
func prob_extend(q *string, d *[][]float64, index *map[string][]int, params *Params) *[]*Hsp {
	hsps := []*Hsp{}

	for q_idx := 0; q_idx < len(*q) - params.W + 1; q_idx++ {
		seed := (*q)[q_idx : q_idx + params.W]

		for _, d_idx := range (*index)[seed] {
			hsp_score := ungapped_extension.ScoreMiddle(q_idx, d_idx, q, d, params)

			// dummy := ""
			// dummy_S_idx := gapped_extension.SIdx(&params.S)

			// fmt.Println("seed hit acc:", hsp_score/utils.SumProbSeq(d_idx, d_idx+params.W-1, &dummy, d, dummy_S_idx, params.HitThres, params))

			hsp_left, score_left := ungapped_extension.Left(q_idx, d_idx, q, d, params)
			hsp_right, score_right := ungapped_extension.Right(q_idx, d_idx, q, d, params)

			if score_left == math.Inf(-1) && score_right == math.Inf(-1) {
				continue
			}
			if score_left > math.Inf(-1) {
				hsp_score += score_left
			}
			if score_right > math.Inf(-1) {
				hsp_score += score_right
			}

			// E-value of HSP
			e_val := 0.711 * float64(len(*q) * len((*d)[0])) * math.Exp(-1.37 * hsp_score)

			// if hsp_score >= params.HspThres {
			if e_val <= params.EThres {
				var hsp Hsp
				hsp.QIdxLeft = (*hsp_left)[0]
				hsp.QIdxRight = (*hsp_right)[0]
				hsp.DIdxLeft = (*hsp_left)[1]
				hsp.DIdxRight = (*hsp_right)[1]
				hsp.Score = hsp_score
				hsp.EVal = e_val
				hsps = append(hsps, &hsp)

				// hsp_acc := hsp.Score/utils.SumProbSeq(hsp.DIdxLeft, hsp.DIdxRight, &dummy, d, dummy_S_idx, params.HitThres, params)
			}
		}
	}
	return &hsps
}

// prob_extend_gap
//
// Gapped extension. Further extend HSPs using Needleman-Wunsch
// Return list of resulting alignments
//
func prob_extend_gap(q *string, d *[][]float64, hsps *[]*Hsp, params *Params) *[]*Alignment {
	S_idx := gapped_extension.SIdx(&params.S)
	alignments := []*Alignment{}

	for i := 0; i < len(*hsps); i++ {
		alignment := gapped_extension.Extend((*hsps)[i], q, d, S_idx, params)
		alignments = append(alignments, &alignment)
	}
	sort.Slice(alignments, func(i, j int) bool {
		// return alignments[i].Accuracy > alignments[j].Accuracy
		return alignments[i].EVal < alignments[j].EVal
	})
	return &alignments
}